akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala

```diff
@@ -16,177 +20,272 @@ import akka.util.{ ByteString, OptionVal }
+    }
+  }
 
-  // FIXME too specialized...
-  def foldLeftValues[A](zero: A)(f: (A, T) ⇒ A): A = {
-    var acc: A = zero
-    var hit = 0
-    var i = 0
-    while (i < capacity && hit < _usedSlots) {
-      val it = backing(i)
-      if (it != null) {
-        acc = f(acc, it)
-        hit += 1
-      }
-      i += 1
+  def deserialize(inboundEnvelope: InboundEnvelope): Unit = {
+    if (inboundEnvelope.flag(EnvelopeBuffer.MetadataPresentFlag)) {
+      inboundEnvelope.envelopeBuffer.byteBuffer.position(EnvelopeBuffer.MetadataContainerAndLiteralSectionOffset)
+      deserializeRaw(inboundEnvelope)
     }
-    acc
   }
 
-  /** Heavy operation, only used for error logging */
-  def keysWithValues: List[Int] = backing.zipWithIndex.filter({ case (t, id) ⇒ t != null }).map(_._2).toList
-
-  def foreach(f: (Byte, T) ⇒ Unit) = {
-    var i = 0
-    var hit = 0
-    while (i < capacity && hit < _usedSlots) {
-      val t = backing(i)
-      if (t != null) {
-        f(i.toByte, t)
-        hit += 1
+  def deserializeRaw(inboundEnvelope: InboundEnvelope): Unit = {
+    val buffer = inboundEnvelope.envelopeBuffer.byteBuffer
+    val length = buffer.getInt
+    val endPos = buffer.position() + length
+    try {
+      if (instruments.nonEmpty) {
+        var i = 0
+        while (i < instruments.length && buffer.position() < endPos) {
+          val instrument = instruments(i)
+          val startPos = buffer.position()
+          val keyAndLength = buffer.getInt
+          val dataPos = buffer.position()
+          val key = getKey(keyAndLength)
+          val length = getLength(keyAndLength)
+          var nextPos = dataPos + length
+          val identifier = instrument.identifier
+          if (key == identifier) {
+            try {
+              deserializeInstrument(instrument, inboundEnvelope, buffer)
+            } catch {
+              case NonFatal(t) ⇒
+                log.debug(
+                  "Skipping deserialization of RemoteInstrument {} since it failed with {}",
+                  instrument.identifier, t.getMessage)
+            }
+            i += 1
+          } else if (key > identifier) {
+            // since instruments are sorted on both sides skip this local one and retry the serialized one
+            log.debug("Skipping local RemoteInstrument {} that has no matching data in the message", identifier)
+            nextPos = startPos
+            i += 1
+          } else {
+            // since instruments are sorted on both sides skip the serialized one and retry the local one
+            log.debug("Skipping serialized data in message for RemoteInstrument {} that has no local match", key)
+          }
+          buffer.position(nextPos)
+        }
+      } else {
+        log.debug(
+          "Skipping serialized data in message for RemoteInstrument(s) {} that has no local match",
+          remoteInstrumentIdIteratorRaw(buffer, endPos).mkString("[", ", ", "]"))
```

Comment:
You're right. The `mkString` is evaluated directly here. I'll fix it. I wonder why I was under the impression that it was call by name?